### PR TITLE
Edit worker

### PR DESCRIPTION
**Pull request by JohnFarrellDev**
## Link to JIRA ticket

Two relevant tickets for this PR.

[Updating Workers](https://hackney.atlassian.net/jira/software/projects/SCS/boards/51?selectedIssue=SCS-656)
[Deactivating Workers](https://hackney.atlassian.net/jira/software/projects/SCS/boards/51?selectedIssue=SCS-657)

### *What is the problem we're trying to solve*

We want managers to be able to edit a worker information including:

Worker name
Worker email
Worker context flag (if they work with adults or children)
Worker role
Worker teams
The the worker is active and can therefore be allocated, when deactivating a worker they cannot have any current allocations

Changes to a worker should be tracked so their is an audit log, audit log does not need to be REST accessible

### *What changes have we introduced*

Created a new PATCH endpoint to allow for the updating of an existing Worker

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Are there any next steps that need to addressed after merging this PR? Add them here.

### CODE REVIEW

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateWorkerRequestTests.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                (TestHelpers.CreateUpdateWorkersRequest(contextFlag: ""), "Context flag must be provided\nContext flag must be 'A' or 'C'"),
+                (TestHelpers.CreateUpdateWorkersRequest(contextFlag: _faker.Random.String2(2, "AC")), "Context flag must be no longer than 1 character"),
+                (TestHelpers.CreateUpdateWorkersRequest(contextFlag: "B"), "Context flag must be 'A' or 'C'"),
+                (TestHelpers.CreateUpdateWorkersRequest(role: ""), "Role must be provided"),
+                (TestHelpers.CreateUpdateWorkersRequest(role: _faker.Random.String2(201)), "Role provided is too long (more than 200 characters)"),
+                (TestHelpers.CreateUpdateWorkersRequest(dateStart: DateTime.Now.AddMinutes(1)), "Date cannot be set in the future"),
+                (TestHelpers.CreateUpdateWorkersRequest(createATeam: false), "A team must be provided"),
+                (TestHelpers.CreateUpdateWorkersRequest(teamName: _faker.Random.String2(201)), "Team name must be no more than 200 characters"),
+                (TestHelpers.CreateUpdateWorkersRequest(teamId: 0), "Team ID must be greater than 0")
+            };
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateWorkerRequestTests.cs:47]: Another way of doing this is to use `TestCase`, e.g. https://github.com/LBHackney-IT/social-care-case-viewer-api/blob/0c6ea7367fda654f2da94ce451b5b16bac3e1703/SocialCareCaseViewerApi.Tests/V1/Boundary/Request/ListTeamsRequestTests.cs#L52-L61.

File: SocialCareCaseViewerApi.Tests/V1/Controllers/WorkerControllerTests.cs [hunk 2]
```diff
@@ -86,6 +86,57 @@ public void CreateWorkerReturns422StatusWhenCreateWorkerExceptionThrown()
             response.Value.Should().Be(errorMessage);
         }
 
+        [Test]
+        public void UpdateWorkerReturns201StatusAndWorkerWhenSuccessful()
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/Controllers/WorkerControllerTests.cs:90]:
```suggestion
        public void UpdateWorkerReturns204StatusAndWorkerWhenSuccessful()
```

File: SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs [hunk 3]
```diff
@@ -210,8 +208,120 @@ public void CreateWorkerWithInvalidTeamIdThrowsPostWorkerException()
...
+
+            var request = TestHelpers.CreateUpdateWorkersRequest(teamId: team.Id, workerId: worker.Id, isActive: false);
+            _classUnderTest.UpdateWorker(request);
+
+            var updatedWorker = DatabaseContext.Workers.First(w => w.Id == worker.Id);
+            updatedWorker.DateStart.Should().BeNull();
+            updatedWorker.DateEnd.Should().NotBeNull();
+        }
+
+        private static void CompareUpdatedWorkerAndRequest(Worker updatedWorker, UpdateWorkerRequest request,
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs:306]: Suggested renaming to so when we read it in the test, it reads like an expectation.

```suggestion
        private static void UpdatedWorkerShouldBeEquivalentToRequest(Worker updatedWorker, UpdateWorkerRequest request,
```

File: SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs [hunk 4]
```diff
@@ -210,8 +208,120 @@ public void CreateWorkerWithInvalidTeamIdThrowsPostWorkerException()
...
+
+            var updatedWorker = DatabaseContext.Workers.First(w => w.Id == worker.Id);
+
+            updatedWorker.Should().NotBeEquivalentTo(originalWorker);
+
+            CompareUpdatedWorkerAndRequest(updatedWorker, request, worker, team);
+        }
+
+        [Test]
+        public void UpdateWorkerIsActiveFlipsDateStartAndDateEndFromTimeToNull()
```
Comment (chubberlisk) [inline SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs:282]: Not close to how workers work but just a thought: is it possible to make a worker go from `isActive: false -> isActive: true`?

**Review by chubberlisk — APPROVED** [review]
Looks good to me! Just some suggestions on the readability of tests.

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateWorkerRequestTests.cs (hunk 1 above)
Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateWorkerRequestTests.cs:47]: I think [TestCase()] only works with the primitive data types and not objects.

I could use [TestCaseSource](https://docs.nunit.org/articles/nunit/writing-tests/attributes/testcasesource.html) I think, I'm happy to implement this if you think it is tidier.

File: SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs (hunk 4 above)
Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs]: I added a test for worker going from `false to true` and `true to false` for `isActive`

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateWorkerRequestTests.cs (hunk 1 above)
Comment (chubberlisk) [reply SocialCareCaseViewerApi.Tests/V1/Boundary/Request/UpdateWorkerRequestTests.cs:47]: Ahhhhh, that's a shame! Never mind - it's okay as is.

**Pull request merged**